spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,124 @@
+
+	private final ProtobufMessageConverter protobufMessageConverter;
+
+	private final Class<? extends com.google.protobuf.Message> defaultType;
+
+	private ClassLoader beanClassLoader;
+
+	private Expression typeIdExpression = new FunctionExpression<Message<?>>(
+			(message) -> message.getHeaders().get(ProtoHeaders.TYPE));
+
+	private EvaluationContext evaluationContext;
+
+	/**
+	 * Construct an instance with the supplied default type to create.
+	 * @param defaultType the type.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType) {
+		this(defaultType, new ProtobufMessageConverter());
+	}
+
+	/**
+	 * Construct an instance with the supplied default type and ProtobufMessageConverter instance.
+	 * @param defaultType the type.
+	 * @param protobufMessageConverter the message converter used.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType,
+			ProtobufMessageConverter protobufMessageConverter) {
+		Assert.notNull(defaultType, "'defaultType' must not be null");
+		Assert.notNull(protobufMessageConverter, "'protobufMessageConverter' must not be null");
+		this.defaultType = defaultType;
+		this.protobufMessageConverter = protobufMessageConverter;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpression(Expression expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+		this.typeIdExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpressionString(String expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+
+		Class<? extends com.google.protobuf.Message> targetClass = null;
+		Object value = this.typeIdExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			targetClass = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				targetClass = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (targetClass == null) {
+			this.logger.trace("Empty SpEL type expression. Fallback to the defaultType!");
```

Comment:
How about this:
```suggestion
			this.logger.trace(() -> "The 'typeIdExpression' (" + this.typeIdExpression + ") returned 'null' for: " + message + ". Falling back to the defaultType: " + this.defaultType);
```
?
I mean to give as much info to end-user as possible to let them to understand the reason behind behavior.

Reply:
More descriptive trace message is better, but I start to think that we should remove it all together. Here is why: 

The `typeIdExpression` and the `proto_type` msg header only make sence when the message was encoded via the `ToProtobufTransofmer`. But the `FromProbufTransformer` can be used to de-ser protobuf content creted with custom `ProtobufMessageConverter` implementations, 3rd party tools and even different programming language. In this case the `defaultType` will be the normal way to configure and use the `FromProtobufTransoformer`. Logging trace messges in this case would be incorrect, anoying and could affect the performance.